### ISSUE TITLE
DateSubject

### ISSUE THREAD

**Issue by ALWAL12** [google/truth#420]
I see that you added subjects from JDK 8, but I wonder if you planned on adding `java.time` classes such as `LocalDate`, `LocalTime`, `LocalDateTime`, `ZonedDateTime` and `Instant`. 

I am willing to help on this. I'm not a pro with dates and such, so I might need some help with the API of these subjects.

It would be useful to compare dates, to know if it's the same date given same/different timezones, etc.

**Comment by kluever** [comment]
We have an internal bug about this, but we've yet to write up an API proposal. Here's some snippets from the internal bug:

We briefly discussed subjects for JodaTime, but maybe we want to re-evaluate for java.time types? I'm imaging something like:
```java
  assertThat(duration).isEqualToMinutes(1);
  assertThat(duration).isEqualToSeconds(60);
  assertThat(duration).isEqualToMillis(60000);
```

Those otherwise would have to be:
```java
  assertThat(duration).isEqualTo(Duration.ofMinutes(1));
  assertThat(duration).isEqualTo(Duration.ofSeconds(60));
  assertThat(duration).isEqualTo(Duration.ofMillis(60000));
```

Someone else also asked for something like:
```java
assertThat(actualInstant).isWithin(Duration.ofSeconds(5)).of(expectedInstant);
```

@ALWAL12 Did you have a specific use or API in mind? Since most of the `java.time` types implement `Comparable`, you will automatically get things like `isGreaterThan()`, `isLessThan()`, etc.

**Comment by ALWAL12** [comment]
I believe special cases happen when you work with timezones. I understand most date objects are `Comparable` and unaware of timezones, but `ZonedDateTime` are apart. Assume this test:

```java
ZonedDateTime montreal = ZonedDateTime.of(LocalDate.of(2012, Month.JANUARY, 28), LocalTime.of(16, 0), ZoneId.of("America/Montreal"));
ZonedDateTime vancouver = ZonedDateTime.of(LocalDate.of(2012, Month.JANUARY, 28), LocalTime.of(13, 0), ZoneId.of("America/Vancouver")); // 3 hours apart

assertThat(montreal).isEqualTo(vancouver);
```
They aren't technically equals due to different timezones, but they represent the same point in time. We have many cases that we receive a zoned date time but we returns back UTC. We always need to convert our `ZonedDateTime` to `Instant` to make those tests pass.

They are many different ways to represent an API, such as 
`assertThat(montreal).hasSameInstantOf(vancouver);`
or
`assertThat(montreal).isEqualTo(vancouver).inSameTimezone();`
or
`assertThat(montreal).inTimezone(ZoneOffset.UTC).isEqualTo(vancouver);`

**Comment by kluever** [comment]
So without the subject, you'd have to write:
  `assertThat(montreal.toInstant().isEqualTo(vancouver.toInstant());`
right?

**Comment by ALWAL12** [comment]
Exactly. That would be kind of a shortcut. In our case, we also mix these dates with `Optional` so the line is a bit messier than this. 😅 

I can't think of other cases at the moment, but there is room for improvements or more-natural assertions about dates.

**Comment by vorburger** [comment]
> I assume the precision truncation is for allowing some wiggle room when comparing now-values? If you inject an `InstantSource` into your code, you'd be able to precisely control the value of "now", and thus you probably wouldn't need that wiggle room.

This in the context of asserting something while testing a Jackson YAML writer and reader, where the (I think) `com.fasterxml.jackson.datatype.jsr310.JavaTimeModule` looses precision, because YAML represents timestamps as decimal (seconds.nanoseconds), which can lose a few nanoseconds of precision, due to floating-point representation limits. Injecting an `InstantSource` is "impractical", so I wanted to handle it in the test, and that's how I found this (ancient) issue.

> Anyways, for flexibility, instead of `isEqualToMillis(Instant)`, you might consider something like `isEqualTo(Instant, ChronoUnit)`. 

I like your idea for the new signature - it's more "open", and still very readable! So my utility is now:

```java
package dev.enola.common.testlib;

public final class InstantSubject extends Subject {

    public static InstantSubject assertThat(@Nullable Instant actual) {
        return assertAbout(InstantSubject::new).that(actual);
    }

    @Nullable private final Instant actual;

    private InstantSubject(FailureMetadata metadata, @Nullable Instant actual) {
        super(metadata, actual);
        this.actual = actual;
    }

    public void isEqualTo(Instant expected, ChronoUnit precision) {
        if (actual == null) {
            failWithActual(simpleFact("expected not to be null"));
            return;
        }
        check("truncatedTo(%s)", precision)
                .that(actual.truncatedTo(precision))
                .isEqualTo(expected.truncatedTo(precision));
    }
}
```

I'm happy if this is ever useful to anyone else stumbling across this again in the future.
